lib/interactive_viewer.ml

```diff
@@ -115,44 +115,163 @@ let toggle_view_mode () =
+        mine_hunk_length,
+        their_hunk_length )
+  in
+
   match hunks with
-  | [] -> (List.rev mine_acc, List.rev their_acc)
-  | (`Common _ as common) :: t ->
-      split_and_align_hunk t (common :: mine_acc) (common :: their_acc)
-  | `Mine s :: `Their s' :: t ->
-      split_and_align_hunk t (`Mine s :: mine_acc) (`Their s' :: their_acc)
-  | `Mine s :: t ->
-      split_and_align_hunk t (`Mine s :: mine_acc) (`Common "" :: their_acc)
-  | `Their s :: t ->
-      split_and_align_hunk t (`Common "" :: mine_acc) (`Their s :: their_acc)
-
-let lines_to_ui lines attr_change =
-  List.map
-    (fun line ->
-      let content, attr =
-        match line with
-        | `Common s -> (s, Notty.A.empty)
-        | `Mine s -> (s, attr_change)
-        | `Their s -> (s, attr_change)
+  | [] ->
+      ( List.rev mine_acc,
+        List.rev their_acc,
+        mine_hunk_length,
+        their_hunk_length )
+  | _ -> (
+      let ( temp_mine,
+            temp_their,
+            new_mine_hunk_length,
+            new_their_hunk_length,
+            remaining_hunks ) =
+        buffer_changes [] [] mine_hunk_length their_hunk_length hunks
+      in
+      let ( balanced_mine,
+            balanced_their,
+            final_mine_hunk_length,
+            final_their_hunk_length ) =
+        append_balanced temp_mine temp_their new_mine_hunk_length
+          new_their_hunk_length
       in
-      W.string ~attr content)
-    lines
+      let updated_mine_acc = List.rev_append balanced_mine mine_acc in
+      let updated_their_acc = List.rev_append balanced_their their_acc in
+      match remaining_hunks with
+      | `Common line :: t ->
+          let common_mine_acc = Common line :: updated_mine_acc in
+          let common_their_acc = Common line :: updated_their_acc in
+          split_and_align_hunk t common_mine_acc common_their_acc
+            final_mine_hunk_length final_their_hunk_length
+      | _ ->
+          split_and_align_hunk remaining_hunks updated_mine_acc
+            updated_their_acc final_mine_hunk_length final_their_hunk_length)
+
+let lines_with_numbers lines attr_change line_num prefix =
+  List.fold_left
+    (fun acc line ->
+      match line with
+      | Common s ->
+          line_num := !line_num + 1;
+          let content = Printf.sprintf "%3d   %s" !line_num s in
+          (content, Notty.A.empty) :: acc
+      | Change (Mine, s) ->
+          incr line_num;
+          let content = Printf.sprintf "%3d %s %s" !line_num prefix s in
+          (content, attr_change) :: acc
+      | Change (Their, s) ->
+          incr line_num;
+          let content = Printf.sprintf "%3d %s %s" !line_num prefix s in
+          (content, attr_change) :: acc
+      | Empty ->
+          let content = Printf.sprintf "      " in
+          (content, Notty.A.empty) :: acc)
+    [] lines
+  |> List.rev
+  |> List.map (fun (content, attr) -> W.string ~attr content)
+
+let create_summary start_line_num hunk_length attr change_type =
```

Comment:
Because for some reason using `Option.to_list` works but not a normal list. I need to make the types match up. 
```
  let space = Ui.space 1 0 in
  Ui.hcat
    [
      Ui.resize ~w:0 ~sw:2
        (Ui.vcat (Option.to_list summary_mine @ content_mine));
      space;
      Ui.resize ~w:0 ~sw:2
        (Ui.vcat (Option.to_list summary_their @ content_their));
    ]
```

Reply:
What you want is probably to add one element at the beginning of a list:
```
  summary_mine :: content_mine
```